### PR TITLE
:seedling: add lint check to ensure that all exported funcs are properly documented

### PR DESCRIPTION
**Pull request by camilamacedo86**
**Description**
All exported funcs should be doc so the lint will ensure that.

### CODE REVIEW

File: test/e2e/utils/util.go [hunk 1]
```diff
@@ -25,6 +25,7 @@ import (
 )
 
 const (
+	// KubebuilderBinName define the name of the kubebuilder binary tto be used in the tests
```
Comment (gabbifish) [inline test/e2e/utils/util.go:28]: nit: 
```suggestion
	// KubebuilderBinName define the name of the kubebuilder binary to be used in the tests
```

**Review by gabbifish — APPROVED** [review]
/approve

This looks great--just added one nit related to a small typo :)

Let's also make sure we can get the 1-14 e2e test to pass.

**Comment by camilamacedo86** [comment]
Hi @gabbifish, 

Really tks for your review. The nit is fixed now 👍

**Comment by gabbifish** [comment]
/approve

awesome!

File: .golangci.yml [hunk 2]
```diff
@@ -1,3 +1,15 @@
+issues:
+  # don't skip warning about doc comments
+  # don't exclude the default set of lint
+  exclude-use-default: false
+  # restore some of the defaults
+  # (fill in the rest as needed)
+  exclude-rules:
+    - linters: [errcheck]
+      text: "Error return value of .((os\\.)?std(out|err)\\..*|.*Close|.*Flush|os\\.Remove(All)?|.*printf?|os\\.(Un)?Setenv). is not checked"
```
Comment (estroz) [inline .golangci.yml:9]: Where is this regexp coming from?

File: pkg/cli/cli_suite_test.go [hunk 3]
```diff
@@ -77,8 +77,13 @@ type mockInitPlugin struct{ mockPlugin }
 type mockCreateAPIPlugin struct{ mockPlugin }
 type mockCreateWebhookPlugin struct{ mockPlugin }
 
-func (p mockInitPlugin) GetInitPlugin() plugin.Init                            { return p }
-func (p mockCreateAPIPlugin) GetCreateAPIPlugin() plugin.CreateAPI             { return p }
+// GetInitPlugin will return the plugin which is responsible for initialized the project
+func (p mockInitPlugin) GetInitPlugin() plugin.Init { return p }
+
+// GetCreateAPIPlugin will return the plugin which is responsible for scaffold apis for the project
```
Comment (estroz) [inline pkg/cli/cli_suite_test.go:83]:
```suggestion
// GetCreateAPIPlugin will return the plugin which is responsible for scaffolding APIs for the project
```

File: pkg/cli/cli_suite_test.go [hunk 4]
```diff
@@ -77,8 +77,13 @@ type mockInitPlugin struct{ mockPlugin }
...
+func (p mockCreateAPIPlugin) GetCreateAPIPlugin() plugin.CreateAPI { return p }
+
+// GetCreateWebhookPlugin will return the plugin which is responsible for scaffold webhooks for the project
```
Comment (estroz) [inline pkg/cli/cli_suite_test.go:86]:
```suggestion
// GetCreateWebhookPlugin will return the plugin which is responsible for scaffolding webhooks for the project
```

File: pkg/model/resource/resource.go [hunk 5]
```diff
@@ -66,6 +66,7 @@ func wrapKey(key string) string {
 	return fmt.Sprintf("%%[%s]", key)
 }
 
+// Replacer replaces the key values for the resources fragments
```
Comment (estroz) [inline pkg/model/resource/resource.go:69]:
```suggestion
// Replacer returns a strings.Replacer that replaces resource keywords with values.
```

File: pkg/plugin/interfaces.go [hunk 6]
```diff
@@ -22,6 +22,7 @@ import (
 	"sigs.k8s.io/kubebuilder/pkg/model/config"
 )
 
+// Base defines the common base for the plugins
```
Comment (estroz) [inline pkg/plugin/interfaces.go:25]:
```suggestion
// Base defines the common base for all plugins
```

File: pkg/plugin/interfaces.go [hunk 7]
```diff
@@ -36,11 +37,13 @@ type Base interface {
 	SupportedProjectVersions() []string
 }
 
+// Deprecated represents the deprecations messages that should be implemented for an plugin
```
Comment (estroz) [inline pkg/plugin/interfaces.go:40]:
```suggestion
// Deprecated is implemented by plugins that are deprecated.
```

File: pkg/plugin/interfaces.go [hunk 8]
```diff
@@ -36,11 +37,13 @@ type Base interface {
...
 type Deprecated interface {
 	// DeprecationWarning returns a string indicating a plugin is deprecated.
 	DeprecationWarning() string
 }
 
+// GenericSubcommand represents the plugins subcommands
```
Comment (estroz) [inline pkg/plugin/interfaces.go:46]:
```suggestion
// GenericSubcommand contains plugin operation methods.
```

File: pkg/plugin/interfaces.go [hunk 9]
```diff
@@ -56,6 +59,7 @@ type GenericSubcommand interface {
 	InjectConfig(*config.Config)
 }
 
+// Context represent the plugins contexts
```
Comment (estroz) [inline pkg/plugin/interfaces.go:62]:
```suggestion
// Context is the runtime context for a plugin.
```

File: pkg/plugin/interfaces.go [hunk 10]
```diff
@@ -66,32 +70,38 @@ type Context struct {
 	Examples string
 }
 
+// InitPluginGetter represents the Init plugin Getters
```
Comment (estroz) [inline pkg/plugin/interfaces.go:73]:
```suggestion
// InitPluginGetter gets an Init plugin
```

File: pkg/plugin/interfaces.go [hunk 11]
```diff
@@ -66,32 +70,38 @@ type Context struct {
...
 type InitPluginGetter interface {
 	Base
 	// GetInitPlugin returns the underlying Init interface.
 	GetInitPlugin() Init
 }
 
+// Init represents the init commands
```
Comment (estroz) [inline pkg/plugin/interfaces.go:80]:
```suggestion
// Init represents an `init` command
```

File: pkg/plugin/interfaces.go [hunk 12]
```diff
@@ -66,32 +70,38 @@ type Context struct {
...
 type Init interface {
 	GenericSubcommand
 }
 
+// CreateAPIPluginGetter represents the create api plugin Getters
```
Comment (estroz) [inline pkg/plugin/interfaces.go:85]:
```suggestion
// CreateAPIPluginGetter gets a CreateAPI plugin
```

File: pkg/plugin/interfaces.go [hunk 13]
```diff
@@ -66,32 +70,38 @@ type Context struct {
...
 	// GetCreateAPIPlugin returns the underlying CreateAPI interface.
 	GetCreateAPIPlugin() CreateAPI
 }
 
+// CreateAPI represents the create api sub commands
 type CreateAPI interface {
 	GenericSubcommand
 }
 
+// CreateWebhookPluginGetter defines the getter interface for CreateWebhook
```
Comment (estroz) [inline pkg/plugin/interfaces.go:97]:
```suggestion
// CreateWebhookPluginGetter gets a CreateWebhook plugin
```

File: pkg/plugin/interfaces.go [hunk 14]
```diff
@@ -66,32 +70,38 @@ type Context struct {
...
 }
 
+// CreateAPIPluginGetter represents the create api plugin Getters
 type CreateAPIPluginGetter interface {
 	Base
 	// GetCreateAPIPlugin returns the underlying CreateAPI interface.
 	GetCreateAPIPlugin() CreateAPI
 }
 
+// CreateAPI represents the create api sub commands
```
Comment (estroz) [inline pkg/plugin/interfaces.go:92]:
```suggestion
// CreateAPI represents a `create api` command
```

File: pkg/plugin/interfaces.go [hunk 15]
```diff
@@ -66,32 +70,38 @@ type Context struct {
...
 }
 
+// CreateWebhookPluginGetter defines the getter interface for CreateWebhook
 type CreateWebhookPluginGetter interface {
 	Base
 	// GetCreateWebhookPlugin returns the underlying CreateWebhook interface.
 	GetCreateWebhookPlugin() CreateWebhook
 }
 
+// CreateWebhook represents the create webhook subcommands
```
Comment (estroz) [inline pkg/plugin/interfaces.go:104]:
```suggestion
// CreateWebhook represents a `create webhook` command
```

File: pkg/plugin/v2/plugin.go [hunk 16]
```diff
@@ -35,15 +35,29 @@ var (
 	_ plugin.CreateWebhookPluginGetter = Plugin{}
 )
 
+// Plugin defines the Plugin. e.g go.kubebuilder.io/v3-alpha
```
Comment (estroz) [inline pkg/plugin/v2/plugin.go:38]:
```suggestion
// Plugin wraps Init, CreateAPI, and CreateWebhook plugins for this plugin version.
```

File: pkg/plugin/v3/plugin.go [hunk 17]
```diff
@@ -35,15 +35,29 @@ var (
 	_ plugin.CreateWebhookPluginGetter = Plugin{}
 )
 
+// Plugin defines the Plugin. e.g go.kubebuilder.io/v3-alpha
```
Comment (estroz) [inline pkg/plugin/v3/plugin.go:38]:
```suggestion
// Plugin wraps Init, CreateAPI, and CreateWebhook plugins for this plugin version.
```

File: pkg/plugin/v2/plugin.go [hunk 18]
```diff
@@ -35,15 +35,29 @@ var (
...
 	createAPIPlugin
 	createWebhookPlugin
 }
 
-func (Plugin) Name() string                                   { return pluginName }
-func (Plugin) Version() plugin.Version                        { return pluginVersion }
-func (Plugin) SupportedProjectVersions() []string             { return supportedProjectVersions }
-func (p Plugin) GetInitPlugin() plugin.Init                   { return &p.initPlugin }
-func (p Plugin) GetCreateAPIPlugin() plugin.CreateAPI         { return &p.createAPIPlugin }
+// Name returns the name of the plugin. e.g go.kubebuilder.io
```
Comment (estroz) [inline pkg/plugin/v2/plugin.go:45]: I don't think comments are required on concrete methods of an interface, and if they exist
can be more confusing/duplicative. I'd prefer these removed if they are not required to pass the linter.

File: pkg/plugin/v3/plugin.go [hunk 19]
```diff
@@ -35,15 +35,29 @@ var (
...
 	createAPIPlugin
 	createWebhookPlugin
 }
 
-func (Plugin) Name() string                                   { return pluginName }
-func (Plugin) Version() plugin.Version                        { return pluginVersion }
-func (Plugin) SupportedProjectVersions() []string             { return supportedProjectVersions }
-func (p Plugin) GetInitPlugin() plugin.Init                   { return &p.initPlugin }
-func (p Plugin) GetCreateAPIPlugin() plugin.CreateAPI         { return &p.createAPIPlugin }
+// Name returns the name of the plugin. e.g go.kubebuilder.io
```
Comment (estroz) [inline pkg/plugin/v3/plugin.go:45]: Ditto above.

**Review by estroz — CHANGES_REQUESTED** [review]
Mostly grammar nits.

File: .golangci.yml (hunk 2 above)
Comment (camilamacedo86) [reply .golangci.yml:9]: we can remove. Maybe is better.

File: pkg/plugin/interfaces.go (hunk 7 above)
Comment (camilamacedo86) [reply pkg/plugin/interfaces.go:40]: I improved that based on your feedback.

File: pkg/plugin/interfaces.go (hunk 8 above)
Comment (camilamacedo86) [reply pkg/plugin/interfaces.go:46]: I improved that based on your feedback.

File: pkg/plugin/interfaces.go (hunk 10 above)
Comment (camilamacedo86) [reply pkg/plugin/interfaces.go:73]: I improved that based on your feedback.

File: pkg/plugin/interfaces.go (hunk 11 above)
Comment (camilamacedo86) [reply pkg/plugin/interfaces.go:80]: I improved that based on your feedback.

File: pkg/plugin/interfaces.go (hunk 12 above)
Comment (camilamacedo86) [reply pkg/plugin/interfaces.go:85]: I improved that based on your feedback.

File: pkg/plugin/interfaces.go (hunk 14 above)
Comment (camilamacedo86) [reply pkg/plugin/interfaces.go:92]: I improved that based on your feedback.

File: pkg/plugin/interfaces.go (hunk 13 above)
Comment (camilamacedo86) [reply pkg/plugin/interfaces.go:97]: I improved that based on your feedback.

File: pkg/plugin/interfaces.go (hunk 15 above)
Comment (camilamacedo86) [reply pkg/plugin/interfaces.go:104]: I improved that based on your feedback.

File: pkg/plugin/v2/plugin.go (hunk 16 above)
Comment (camilamacedo86) [reply pkg/plugin/v2/plugin.go:38]: I improved that based on your feedback.

File: pkg/plugin/v2/plugin.go (hunk 18 above)
Comment (camilamacedo86) [reply pkg/plugin/v2/plugin.go:45]: Developers are able to use the interface to create their own plugins as use the plugins implementations. In this way, IHMO both needs to be properly documented. However, I approved it as well based in your comment.

Comment (camilamacedo86) [reply pkg/plugin/v2/plugin.go:45]: Developers are able to use the interface to create their own plugins as use the plugins implementations. In this way, IHMO both needs to be properly documented. However, I approved it as well based in your comment.

File: pkg/plugin/v3/plugin.go (hunk 17 above)
Comment (camilamacedo86) [reply pkg/plugin/v3/plugin.go:38]: I improved that based on your feedback.

File: pkg/plugin/v2/plugin.go (hunk 18 above)
Comment (estroz) [reply pkg/plugin/v2/plugin.go]: The interface already describes the method's contract. If there are any special things a particular concrete method does, they can be added in a comment; none of the comments being added give any extra relevant information so they aren't helpful and can drift easily. I'd prefer if you removed them, but I won't block the PR on this.

**Pull request merged**